proxy/agent.go

```diff
@@ -0,0 +1,198 @@
+func NewAgent(run func(interface{}, int) error, cleanup func(int),
+	initialRetryBudget int, initialRetryDelay time.Duration) Agent {
+	return &agent{
+		run:                run,
+		cleanup:            cleanup,
+		epochs:             make(map[int]interface{}),
+		config:             make(chan interface{}),
+		status:             make(chan exitStatus),
+		initialRetryBudget: initialRetryBudget,
+		initialRetryDelay:  initialRetryDelay,
+	}
+}
+
+type agent struct {
+	// proxy start-up command
+	run func(interface{}, int) error
+
+	// proxy cleanup command
+	cleanup func(int)
+
+	// desired configuration state
+	desired interface{}
+
+	// active epochs and their configurations
+	epochs map[int]interface{}
+
+	// current configuration is the highest epoch configuration
+	current interface{}
+
+	// channel for posting desired configurations
+	config chan interface{}
+
+	// channel for proxy exit notifications
+	status chan exitStatus
+
+	// retry budget count
+	retryBudget int
+
+	// total budget available
+	initialRetryBudget int
+	initialRetryDelay  time.Duration
+}
+
+type exitStatus struct {
+	epoch int
+	err   error
+}
+
+func (a *agent) Apply(config interface{}) {
+	a.config <- config
+}
+
+// Run executes the reloads until stop channel closes
+func (a *agent) Run(stop <-chan struct{}) {
+	glog.V(2).Info("Starting proxy agent")
+
+	// Throttle processing up to smoothed 1 qps with bursts up to 10 qps
+	rateLimiter := flowcontrol.NewTokenBucketRateLimiter(float32(1), 10)
+
+	// Set default delay to a long duration - reconciliation is a no-op in the regular case.
+	// For permanent errors, reconciliation is still attempted once after the default delay.
+	defaultDelay := 1 * time.Hour
+	delay := defaultDelay
+
+	for {
+		rateLimiter.Accept()
+
+		select {
+		case config := <-a.config:
+			// reset retry budget if and only if the desired config changes
+			if !reflect.DeepEqual(a.desired, config) {
+				a.retryBudget = a.initialRetryBudget
+				delay = defaultDelay
+				a.desired = config
+				a.reconcile()
+			}
+
+		case status := <-a.status:
+			if status.err != nil {
+				glog.V(2).Infof("Epoch %d terminated with an error: %v", status.epoch, status.err)
+			} else {
+				glog.V(2).Infof("Epoch %d exited normally", status.epoch)
+			}
+
+			// cleanup for the epoch
+			a.cleanup(status.epoch)
+
+			// delete epoch record and update current config
+			delete(a.epochs, status.epoch)
+			a.current = a.epochs[a.latestEpoch()]
+
+			// schedule a retry for a transient error
+			if status.err != nil && !reflect.DeepEqual(a.desired, a.current) {
+				if a.retryBudget > 0 {
+					delay = a.initialRetryDelay * (2 << uint(a.initialRetryBudget-a.retryBudget))
+					a.retryBudget = a.retryBudget - 1
+					glog.V(2).Infof("Schedule retry after %v (budget %d)", delay, a.retryBudget)
+				} else {
+					glog.Warningf("Permanent error: budget exhausted trying to fulfill the desired configuration")
+					delay = defaultDelay
```

Comment:
So what do we do in this case? Thinking out aloud, it depends on the error. what is the point of running the agent if we know that we couldn't apply the config? For all we know, we could have N instances of envoy running or something failing. Or the error may resolve itself.